File: accel_integration.py
from typing import Optional, Any

import pandas as pd
import numpy as np
from scipy.integrate import simpson, cumulative_simpson, solve_ivp
from scipy.interpolate import interp1d
import pathlib
import glob
import matplotlib
import matplotlib.pyplot as plt

# Load data
source_folder = pathlib.Path(r"D:\Reverse Telescope Test\accel\Session_2025-10-29_163326")
csv_files = glob.glob(str(source_folder / "*.csv"))
try :
    df_list = [pd.read_csv(f) for f in csv_files[0:50]] # 50 files will eat up about 8 gigs of ram
    df = pd.concat(df_list, ignore_index=True)
except MemoryError as e:
    print("Memory error, you may have too much data.")
    print(e)
    if len(csv_files) > 5 :
        df_list = [pd.read_csv(f) for f in csv_files[0:5]]
    else:
        df_list = [pd.read_csv(csv_files[0])]
    df = pd.read_csv(pathlib.Path.joinpath(source_folder, "*.csv"))
except Exception as e :
    print(e)


t = df['RelativeTime_s'].values
ax = df['Mirror_X_g'].values * 10 # accel output from matlab is in g's. g = 386.1 inches/sec^2 For now just get raw V
ay = df['Mirror_Y_g'].values * 10
az = df['Mirror_Z_g'].values * 10
aarm = df['Desk_Y_g'].values * 10

def plotaccels(
        t: np.ndarray[Any, np.dtype[np.floating[Any]]],
        a: np.ndarray[Any, np.dtype[np.floating[Any]]],
        axis: str,
        suppressPosition: Optional = False,
        suppressVelocity: Optional = False) -> tuple[np.ndarray, Optional[np.ndarray]]:
    if len(t) != len(a):
        raise ValueError("All input arrays must have the same length.")

    plt.figure(figsize=(10, 6))
    v=None
    p=None
    if suppressPosition and suppressVelocity :
        mean = a.mean()
        std = a.std()
        plt.plot(t, a, label=f'{axis} Acceleration (V) (mean={mean:.2f}, std={std:.2f})')
    else:
        plt.plot(t, a, label=f'{axis} Acceleration (V)')
    if not suppressVelocity :
        v = cumulative_simpson(y=a, x=t, initial=0)
        plt.plot(t, v, label=f'{axis} Velocity (in/s)')
    if not suppressPosition:
        p = cumulative_simpson(y=v, x=t, initial=0)
        plt.plot(t, p, label=f'{axis} Displacement (in)')
    plt.legend()
    plt.xlabel('Time (s)')
    plt.ylabel('Value')
    plt.title(f'{axis} Acceleration → Velocity → Displacement')
    plt.show()

    return v, p

vx, px = plotaccels(t, ax, "X", suppressPosition=True, suppressVelocity=True)
vy, py = plotaccels(t, ay, "Y", suppressPosition=True, suppressVelocity=True)
vz, pz = plotaccels(t, az, "Z", suppressPosition=True, suppressVelocity=True)
varm, parm = plotaccels(t, aarm, "Desk Y", suppressPosition=True, suppressVelocity=True)
#
# # Integration using Runge-Kutta (solve_ivp)
# # Interpolation for acceleration
# ax_fun = interp1d(t, ax, kind='linear', fill_value='extrapolate')
#
# def odefun_x(tt, y):
#     return [ax_fun(tt), y[0]]  # [dv/dt, dx/dt]
#
# y0 = [0, 0]
# sol_x = solve_ivp(odefun_x, [t[0], t[-1]], y0, t_eval=t, method='RK45')
# velocity_x = sol_x.y[0]
# position_x = sol_x.y[1]

File: accel_integration.m
%%% This is a script that will numerically integrate the accelerometer data that Nate has produced.
%%% Keep in mind that accelerometer data looks like this:
%%% AbsoluteTime,RelativeTime_s,Mirror_Y_g,Mirror_X_g,Mirror_Z_g,Desk_Y_g
%%% 2025-10-29 16:33:27.200,0.000000,-0.001054,-0.005731,-0.003451,-0.002179
%%% Code was produced primarily by Microsoft Copilot/Chat GPT-5

% Load data
source_folder = 'D:\Reverse Telescope Test\accel\Session_2025-10-29_163326';
data = readtable(fullfile(source_folder, 'AccelData_2025-10-29_163326_File0001.csv'));

% Extract time and acceleration (convert g to m/s^2)
t = data.RelativeTime_s;
ax = data.Mirror_X_g * 9.80665;
ay = data.Mirror_Y_g * 9.80665;
az = data.Mirror_Z_g * 9.80665;

%%Simpson's Rule

% Interpolate acceleration for continuous function
ax_fun = @(tt) interp1(t, ax, tt, 'linear');
ay_fun = @(tt) interp1(t, ay, tt, 'linear');
az_fun = @(tt) interp1(t, az, tt, 'linear');

% Velocity from acceleration
vx = arrayfun(@(tt) integral(ax_fun, t(1), tt), t);
%vy = arrayfun(@(tt) integral(ay_fun, t(1), tt, 'Method', 'Simpson'), t);
%vz = arrayfun(@(tt) integral(az_fun, t(1), tt, 'Method', 'Simpson'), t);

% Position from velocity
px = arrayfun(@(tt) integral(@(tau) interp1(t, vx, tau, 'linear'), t(1), tt), t);
%py = arrayfun(@(tt) integral(@(tau) interp1(t, vy, tau, 'linear'), t(1), tt, 'Method', 'Simpson'), t);
%pz = arrayfun(@(tt) integral(@(tau) interp1(t, vz, tau, 'linear'), t(1), tt, 'Method', 'Simpson'), t);

plot(t,px)

%%Runge-Kutta with ODE45

% Define ODE system: y(1)=velocity, y(2)=position
ax_fun = @(tt) interp1(t, ax, tt, 'linear');
odefun_x = @(tt, y) [ax_fun(tt); y(1)];

% Initial conditions
y0 = [0; 0];

% Solve
[t_sol, y_sol_x] = ode45(odefun_x, [t(1) t(end)], y0);
% y_sol_x(:,1) = velocity, y_sol_x(:,2) = position


